docs/guides/prompt-caching.md

```diff
@@ -0,0 +1,137 @@
+
+# Add a system message with cache control
+chat.with_instructions("You are an AI assistant tasked with analyzing literary works.",
+                      cache_control: true)
+
+# Add a large document with cache control
+chat.ask("Here's the entire text of Pride and Prejudice: [long text...]",
+         with: { cache_control: true })
+
+# Now you can ask questions about the document without reprocessing it
+chat.ask("Analyze the major themes in Pride and Prejudice.")
+```
+
+## Pricing
+
+Prompt caching introduces a different pricing structure:
+
+| Model             | Base Input Tokens | Cache Writes  | Cache Hits   | Output Tokens |
+| ----------------- | ----------------- | ------------- | ------------ | ------------- |
+| Claude 3.7 Sonnet | $3 / MTok         | $3.75 / MTok  | $0.30 / MTok | $15 / MTok    |
+| Claude 3.5 Sonnet | $3 / MTok         | $3.75 / MTok  | $0.30 / MTok | $15 / MTok    |
+| Claude 3.5 Haiku  | $0.80 / MTok      | $1 / MTok     | $0.08 / MTok | $4 / MTok     |
+| Claude 3 Haiku    | $0.25 / MTok      | $0.30 / MTok  | $0.03 / MTok | $1.25 / MTok  |
+| Claude 3 Opus     | $15 / MTok        | $18.75 / MTok | $1.50 / MTok | $75 / MTok    |
+
+Note:
+
+- Cache write tokens are 25% more expensive than base input tokens
+- Cache read tokens are 90% cheaper than base input tokens
+- Regular input and output tokens are priced at standard rates
+
+## Tracking Cache Performance
+
+When using prompt caching, you can track the cache performance using the following fields in the response:
+
+```ruby
+response = chat.ask("What are the main characters in Pride and Prejudice?")
+
+puts "Cache creation tokens: #{response.cache_creation_input_tokens}"
+puts "Cache read tokens: #{response.cache_read_input_tokens}"
+puts "Regular input tokens: #{response.input_tokens}"
+puts "Output tokens: #{response.output_tokens}"
+```
+
+## Cache Limitations
+
+- The minimum cacheable prompt length is:
+  - 1024 tokens for Claude 3.7 Sonnet, Claude 3.5 Sonnet, and Claude 3 Opus
+  - 2048 tokens for Claude 3.5 Haiku and Claude 3 Haiku
+- Shorter prompts cannot be cached, even if marked with `cache_control`
+- The cache has a minimum 5-minute lifetime
+- Cache hits require 100% identical prompt segments
+
+## Best Practices
+
+- Place static content (system instructions, context, examples) at the beginning of your prompt
+- Mark the end of the reusable content for caching using the `cache_control` parameter
+- Use cache breakpoints strategically to separate different cacheable prefix sections
+- Regularly analyze cache hit rates and adjust your strategy as needed
+
+## Example: Document Analysis
+
+```ruby
+# Create a chat with Claude
+chat = RubyLLM.chat(model: 'claude-3-5-sonnet')
+
+# Add system instructions with cache control
+chat.with_instructions("You are an AI assistant tasked with analyzing documents.",
+                      cache_control: true)
+
+# Add a PDF document with cache control
+chat.ask("Please analyze this document:",
```

Comment:
this API seems _fine_ to me. We need to tag the messages with "this is where to cache", so the with option seems like a fine place for that.